### DISCUSSION TITLE
`defineRecipe` is not type-safe

### DISCUSSION THREAD

**Discussion by Newbie012** [chakra-ui/panda#1776] [Ideas]
### Description

When I try to define a CSS recipe using PandaCSS, I do not get errors when using invalid or inappropriate values for properties. For example, borderRadius accepts non-strict tokens and non-existing properties like xxxRadius does not give any error.

### Link to Reproduction

https://play.panda-css.com/AdekKLOMx1

### Steps to reproduce

1. Import defineRecipe from PandaCSS.
2. Define a new recipe using defineRecipe.
3. In the recipe, use an unacceptable value (like a non-strict token) for a CSS property like borderRadius.
4. Also try adding a non-existing CSS property.
5. Confirm that no error appears.

### JS Framework

React (TS)

### Panda CSS Version

0.18.3

### Browser

Google Chrome 119

### Operating System

- [ ] macOS
- [X] Windows
- [ ] Linux

### Additional Information

Recipe example:

```ts
defineRecipe({
    className: "button",
    base: {
        color: "this.color.does.not.exist",
        borderRadius: "asd",
        xxxRadius: "dsa",
    }
})
```

**Comment by astahmer** [comment]
adding those lines in your config file should enable the type-safety

```ts
import type { SystemStyleObject as SystemStyleObjectWithTypes } from "styled-system/types"
declare module '@pandacss/dev' {
  export interface SystemStyleObject extends SystemStyleObjectWithTypes {}
}

```

https://play.panda-css.com/4Bv48dvypG

**Comment by Newbie012** [comment]
Thanks for the response, @astahmer

While it helps avoid typing incorrect properties, it still allows me to write invalid values such as `color: 'nonexistant.color'`.

https://play.panda-css.com/iY6nQ811ot

**Comment by astahmer** [comment]
got it, I think this might not be that simple and we'd have to make a lot of utility functions with generics for the typesafety
we actually also had this idea noted in our internal issue tracker:

```ts
const themeFn = createThemeWithTokens({
  colors: {
    primary: { value: "xxx" }
  }
})
const buttonRecipe = themeFn.defineRecipe({
  base: {
    color: "",
        // ^? "primary" is also suggested here, just like it will be suggested in the app code
  }
})
```
![image](https://github.com/chakra-ui/panda/assets/47224540/ae0df637-a98c-40cd-ac24-81a850f651ce)


PR are definitely welcome if you come up with a good implementation !

**Comment by Newbie012** [comment]
Hi @astahmer, after forking the repo, I was able to achieve it in two ways:

## Module Augmentation

Instead of `interface RecipeConfig` extending the default types, it checks whether a (new) interface called `StyledRecipeConfig` is defined. If it does, it uses it, and only then falls back to the current `RecipeConfig<T>`. This is how I did it:

```ts
// packages/types/recipe.ts

type IsEmptyInterface<T> = keyof T extends never ? true : false;

export interface StyledRecipeConfig {}

type DefaultRecipeConfig<T extends RecipeVariantRecord = RecipeVariantRecord> =
  RecipeDefinition<T> & RecipeConfigMeta;

export type RecipeConfig<T extends RecipeVariantRecord = RecipeVariantRecord> =
    IsEmptyInterface<StyledRecipeConfig> extends true
        ? DefaultRecipeConfig<T>
        : StyledRecipeConfig;

// Previous implementation:
// export interface RecipeConfig<T extends RecipeVariantRecord = RecipeVariantRecord>
//   extends RecipeDefinition<T>,
//     RecipeConfigMeta {}
```

after that, I exported `StyledRecipeConfig` so I could apply declaration merging (maybe there's a better way of doing it) by modifying `packages/cli/src/index.ts`:

```ts
import type {
  ...
  StyledRecipeConfig
} from '@panda/types'

...

export type {
  ...
  StyledRecipeConfig
}
```

and then add these lines to the codebase:

```ts
declare module '@pandacss/dev' {
  export interface StyledRecipeConfig<T extends RecipeVariantRecord = RecipeVariantRecord>
    extends CodegenRecipeConfig<T> {}
}
```

And....

![image](https://github.com/chakra-ui/panda/assets/10504365/6094a5ab-f991-4190-89e4-d54d8ba9133f)

🚀🚀🚀

The only issue here is that the error now shows 2 overloads:

![image](https://github.com/chakra-ui/panda/assets/10504365/aa751848-dfc3-407c-ae04-065f0d18e3fe)

Which leads me to the seconds approach:

## Re-declare `defineRecipe` with the codegen types:

```ts
import { RecipeConfig as CodegenRecipeConfig, RecipeVariantRecord } from './styled-system-strict/types'

export function defineSaferRecipe<T extends RecipeVariantRecord>(config: CodegenRecipeConfig<T>): RecipeConfig {
  return config as RecipeConfig
}
```

![image](https://github.com/chakra-ui/panda/assets/10504365/2fc9ffb4-956e-4d6a-8f20-68f799969635)

which now shows a proper error:

![image](https://github.com/chakra-ui/panda/assets/10504365/1ce74326-f507-4eec-bc03-279ab008f92e)

---

If you want to test it, this is what I did: https://github.com/chakra-ui/panda/compare/main...Newbie012:panda:type-safe-define-recipe

**Reply by peterbartels** [reply]
Type safety support for config and slot recipes would be awesome. Would it be possible to make this part of Pandacss by default @astahmer ?

**Reply by astahmer** [reply]
we can't make the above solution (chicken & egg issue, relying on the generated outdir `styled-system` to.. generate itself) part of Panda, feel free to use this snippet tho, it still brings type-safety and is super quick to setup:

```ts
import { RecipeConfig as CodegenRecipeConfig, RecipeVariantRecord } from './styled-system-strict/types'

export function defineSaferRecipe<T extends RecipeVariantRecord>(config: CodegenRecipeConfig<T>): RecipeConfig {
  return config as RecipeConfig
}
```

---

the alternative would be to create a type-safe alternative using the builder pattern, and this is exactly what I've done [here](https://github.com/astahmer/pandabox/tree/main/packages/define-theme)
this approach has different issues: 
- use intensive inferring and type mapping which can lead to slow IDE completion
- can't infer the possible colorPalette values
- can't infer values from presets
- can't infer values for utility when `values` is a function

**Reply by cgatian** [reply]
~Having difficulty making this work with `defineSlotRecipe`.~

Figured it out: 

```ts
export function defineSaferSlotRecipe<
  S extends string = string,
  T extends SlotRecipeVariantRecord<S> = SlotRecipeVariantRecord<S>,
>(config: CodegenSlotRecipeConfig<S, T>): SlotRecipeConfig {
  return config as SlotRecipeConfig;
}
```

**Comment by firzanarmani** [comment]
after observing the createStyleContext commit, wouldn't these utilities be useful to be codegen-ed into the style-system too?

**Comment by Aias** [comment]
The helpers above were life savers once I realized what was happening here.

While I understand the circularity problems of having a recipe generator dependent on the styled system directory it ends up generating, I also would have expected the panda cli to respect the strictness properties of the config, and either warn or error for invalid recipes.

The most pernicious one for me is when setting `shorthands: false` in the config, if you use something like `paddingX` or `paddingY` in a recipe (which people might assume are actually the equivalent longhands for `px` and `py`, then `panda cssgen` will output CSS where `padding-x` and `padding-y` are passed through to the stylesheet rather than remapping to `padding-inline` and `padding-block`. If the shorthands setting was *only* for type safety but the recipes still translated these shorthands on generation, that would make sense to me, but in this case we've got different, invalid output with no warning from the cli.

Is this intended or documented anywhere? I would expect for design system maintainers making heavy use of config recipes to author styles, this could cause some hard-to-diagnose problems.

**Comment by LiahMartens** [comment]
@astahmer for a simpl-ish way of achieving some amount of token type checking in recipes would it be problematic from a theming perspective if I built a theme preset package which just contains the tokens and then in a derivative preset package I use the token preset to generate the styled-system and then usen `css.raw()` to create the styles

ie:
```
import { definePreset, defineRecipe } from "@pandacss/dev";
import { preset as tokens } from 'tokens-preset/preset';
import { css } from '@/styled-system/css'

export const preset = definePreset({
  name: 'preset',
  presets: [tokens],
  theme: {
    recipes: {
      button: defineRecipe({
        className: 'button',
        base: css.raw({
          bg: 'brand.primary'
        })
      })
    },
  }
})
```

Would that cause any underlying problems on initial thoughts?
